### ISSUE TITLE
stats: lots of refreshes over the same table in a short time period

### ISSUE THREAD

**Issue by danhhz** [cockroachdb/cockroach#36027]
Hey, I was running the cdc/initial_scan roachtest to try and figure out why it's been running out of memory and noticed that we ended up with a ton of automatic stats jobs over the same table in a short about of time. Is this expected? These happen to be small tables but if this happened over order_line, for example, it could explain some of my memory issues.

<img width="1319" alt="Screen Shot 2019-03-21 at 9 58 39 AM" src="https://user-images.githubusercontent.com/52528/54770387-55650a80-4bc0-11e9-8e94-2038427c9889.png">

**Comment by rytaft** [comment]
I just merged a PR yesterday to fix this issue -- were you running this on the latest version of master?

**Comment by danhhz** [comment]
It's running at 80fec6280bd515829b11bc89cfe6b0df18a0c265, which does have the pr you mentioned over slack (https://github.com/cockroachdb/cockroach/pull/35871)

**Comment by danhhz** [comment]
Talked to @rytaft offline. Given our heuristics, this is currently expected behavior for these two small tables, so there isn't a worry that this is a bug. It's a bit unfortunate given that the stats for these two tables really don't change that often, but this isn't always true in general. Nothing to do here immediately, really.

**Comment by rytaft** [comment]
cc @awoods187 
With our current model of refreshing after 20% + 500 rows, the `DISTRICT` table, which has 1000 rows in this test, should be refreshed after approximately 700 updates. Given that the `DISTRICT` table is updated in both the payment and new order transactions, and the total throughput is about 300 queries/sec, I'm not surprised that it's being refreshed about twice per minute.

If this still seems too frequent, one option is to increase the base threshold from 500 up to, say, 1000 or 5000. I'm not sure what the right solution is here, but 2 refreshes per minute doesn't seem crazy to me. The bigger problem seems to the admin UI clutter....

**Comment by awoods187** [comment]
I'm going to re-open this because I think it's really bad user experience. 

![image](https://user-images.githubusercontent.com/22278911/54843636-3b542680-4cab-11e9-8841-fef2e9446b0b.png)

I agree that https://github.com/cockroachdb/cockroach/issues/34745 is a potential solution here but i do wonder if there is something smarter we can do here. Like if a table has less than X rows no more frequent than y. Or perhaps as you mentioned the increase to the number of rows changed.

**Comment by awoods187** [comment]
Oh, and even if we filter in the UI, show jobs is now very hard to use:
```
root@localhost:26257/tpcc> show jobs;
        job_id       |     job_type      |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |                           statement                           | user_name |  status   | running_status |             created              |             started              |             finished             |             modified             | fraction_completed |                      error                       | coordinator_id
+--------------------+-------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------+-----------+-----------+----------------+----------------------------------+----------------------------------+----------------------------------+----------------------------------+--------------------+--------------------------------------------------+----------------+
  436494357321154562 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | running   | NULL           | 2019-03-22 18:05:06.506705+00:00 | 2019-03-22 18:05:06.572412+00:00 | NULL                             | 2019-03-22 18:05:06.67362+00:00  |                  1 |                                                  |              2
  436494349332086786 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:05:04.068642+00:00 | 2019-03-22 18:05:04.111668+00:00 | 2019-03-22 18:05:06.476109+00:00 | 2019-03-22 18:05:06.476109+00:00 |                  1 |                                                  |              2
  436494319915401219 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:04:55.091376+00:00 | 2019-03-22 18:04:55.133267+00:00 | 2019-03-22 18:04:56.857413+00:00 | 2019-03-22 18:04:56.857414+00:00 |                  1 |                                                  |              3
  436494317270958083 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:04:54.284355+00:00 | 2019-03-22 18:04:54.420482+00:00 | 2019-03-22 18:04:55.068938+00:00 | 2019-03-22 18:04:55.068939+00:00 |                  1 |                                                  |              3
  436494187388829697 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:04:14.649087+00:00 | 2019-03-22 18:04:14.898816+00:00 | 2019-03-22 18:04:15.60932+00:00  | 2019-03-22 18:04:15.60932+00:00  |                  1 |                                                  |              1
  436494179469885441 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:04:12.232599+00:00 | 2019-03-22 18:04:12.659668+00:00 | 2019-03-22 18:04:14.445942+00:00 | 2019-03-22 18:04:14.445942+00:00 |                  1 |                                                  |              1
  436494141425451010 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:04:00.620564+00:00 | 2019-03-22 18:04:00.828473+00:00 | 2019-03-22 18:04:04.014836+00:00 | 2019-03-22 18:04:04.014837+00:00 |                  1 |                                                  |              2
  436494138763313154 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:03:59.808146+00:00 | 2019-03-22 18:03:59.821617+00:00 | 2019-03-22 18:04:00.466788+00:00 | 2019-03-22 18:04:00.466789+00:00 |                  1 |                                                  |              2
  436494119475773443 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:03:53.922059+00:00 | 2019-03-22 18:03:53.950566+00:00 | 2019-03-22 18:03:54.225199+00:00 | 2019-03-22 18:03:54.2252+00:00   |                  1 |                                                  |              3
  436494114450800643 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:03:52.388553+00:00 | 2019-03-22 18:03:52.518714+00:00 | 2019-03-22 18:03:53.906131+00:00 | 2019-03-22 18:03:53.906131+00:00 |                  1 |                                                  |              3
  436493977657409537 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:03:10.642528+00:00 | 2019-03-22 18:03:10.718855+00:00 | 2019-03-22 18:03:12.211324+00:00 | 2019-03-22 18:03:12.211325+00:00 |                  1 |                                                  |              1
  436493976669257729 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:03:10.341729+00:00 | 2019-03-22 18:03:10.37102+00:00  | 2019-03-22 18:03:10.566278+00:00 | 2019-03-22 18:03:10.566278+00:00 |                  1 |                                                  |              1
  436493938839486466 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:58.796242+00:00 | 2019-03-22 18:02:58.958361+00:00 | 2019-03-22 18:02:59.793323+00:00 | 2019-03-22 18:02:59.793323+00:00 |                  1 |                                                  |              2
  436493931556732930 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:56.573715+00:00 | 2019-03-22 18:02:56.583236+00:00 | 2019-03-22 18:02:58.706102+00:00 | 2019-03-22 18:02:58.706102+00:00 |                  1 |                                                  |              2
  436493915414069251 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:51.64737+00:00  | 2019-03-22 18:02:51.681605+00:00 | 2019-03-22 18:02:52.373204+00:00 | 2019-03-22 18:02:52.373205+00:00 |                  1 |                                                  |              3
  436493909810675715 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:49.937347+00:00 | 2019-03-22 18:02:50.015934+00:00 | 2019-03-22 18:02:51.635736+00:00 | 2019-03-22 18:02:51.635736+00:00 |                  1 |                                                  |              3
  436493774668562433 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:08.695246+00:00 | 2019-03-22 18:02:08.822681+00:00 | 2019-03-22 18:02:10.320376+00:00 | 2019-03-22 18:02:10.320377+00:00 |                  1 |                                                  |              1
  436493772943687681 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:02:08.16961+00:00  | 2019-03-22 18:02:08.193419+00:00 | 2019-03-22 18:02:08.51087+00:00  | 2019-03-22 18:02:08.51087+00:00  |                  1 |                                                  |              1
  436493727737053186 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:01:54.372887+00:00 | 2019-03-22 18:01:54.406803+00:00 | 2019-03-22 18:01:56.561779+00:00 | 2019-03-22 18:01:56.561779+00:00 |                  1 |                                                  |              2
  436493725960372226 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | CREATE STATISTICS __auto__ FROM [61] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:01:53.830693+00:00 | 2019-03-22 18:01:53.904859+00:00 | 2019-03-22 18:01:54.360814+00:00 | 2019-03-22 18:01:54.360814+00:00 |                  1 |                                                  |              2
  436493703505969155 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.district                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | CREATE STATISTICS __auto__ FROM [58] AS OF SYSTEM TIME '-30s' | root      | succeeded | NULL           | 2019-03-22 18:01:46.978147+00:00 | 2019-03-22 18:01:47.088133+00:00 | 2019-03-22 18:01:49.897939+00:00 | 2019-03-22 18:01:49.897939+00:00 |                  1 |                                                  |              3
  436493701463048195 | AUTO CREATE STATS | Table statistics refresh for tpcc.public.warehouse
```

**Comment by rytaft** [comment]
I've just opened PR #36083 which will hopefully address the main UX concerns here.

I think whether or not we increase the minimum rows for a refresh should be a separate discussion. I don't think the fact that we refresh `DISTRICT` twice per minute is necessarily bad, since scanning such a small table twice per minute should have minimal impact on performance. It seems like the main issue is the admin UI clutter and `SHOW JOBS` clutter -- both of which are addressed by the above PR. Comments/ideas welcome.

**Comment by rytaft** [comment]
I believe this can be closed now, since the UX issues were fixed in #36083. Feel free to reopen if not.
